internal/datacoord/channel_manager.go

```diff
@@ -438,3 +521,284 @@ func (c *ChannelManager) findChannel(channelName string) (int64, *channel) {
+
+		err := c.remove(nodeID, chToCleanUp)
+		if err != nil {
+			return fmt.Errorf("failed to remove watch info: %v,%s", chToCleanUp, err.Error())
+		}
+
+		log.Info("channel manager reassign channels", zap.Int64("old node ID", nodeID), zap.Array("updates", updates))
+
+		return c.updateWithTimer(updates, datapb.ChannelWatchState_ToWatch)
+	}
+
+	err := c.remove(nodeID, chToCleanUp)
+	if err != nil {
+		return fmt.Errorf("failed to remove watch info: %v,%s", chToCleanUp, err.Error())
+	}
+
+	if c.dropCallback != nil {
+		log.Debug("try to cleanup removal flag ", zap.String("channel name", channelName))
+		err = c.dropCallback(channelName)
+		if err != nil {
+			return fmt.Errorf("failed to clean up removal flag: %v,%s", chToCleanUp, err.Error())
+		}
+	}
+
+	return nil
+}
+
+type channelStateChecker func(context.Context)
+
+func (c *ChannelManager) watchChannelStatesLoop(ctx context.Context) {
+	defer logutil.LogPanic()
+
+	// REF MEP#7 watchInfo paths are orgnized as: [prefix]/channel/{node_id}/{channel_name}
+	watchPrefix := Params.DataCoordCfg.ChannelWatchSubPath
+	etcdWatcher, timeoutWatcher := c.stateTimer.getWatchers(watchPrefix)
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("watch etcd loop quit")
+			return
+		case ackEvent := <-timeoutWatcher:
+			log.Debug("receive timeout acks from state watcher",
+				zap.Int64("nodeID", ackEvent.nodeID), zap.String("channel name", ackEvent.channelName))
+			c.processAck(ackEvent)
+		case event := <-etcdWatcher:
+			if event.Canceled {
+				log.Warn("watch channel canceled", zap.Error(event.Err()))
+				// https://github.com/etcd-io/etcd/issues/8980
+				if event.Err() == v3rpc.ErrCompacted {
+					go c.watchChannelStatesLoop(ctx)
+					return
+				}
+				// if watch loop return due to event canceled, the datacoord is not functional anymore
+				log.Panic("datacoord is not functional for event canceled")
+			}
+
+			for _, evt := range event.Events {
+				if evt.Type == clientv3.EventTypeDelete {
+					continue
+				}
+				key := string(evt.Kv.Key)
+				watchInfo, err := parseWatchInfo(key, evt.Kv.Value)
+				if err != nil {
+					log.Warn("fail to parse watch info", zap.Error(err))
+					continue
+				}
+
+				// ignore these states
+				state := watchInfo.GetState()
+				if state == datapb.ChannelWatchState_ToWatch ||
+					state == datapb.ChannelWatchState_ToRelease ||
+					state == datapb.ChannelWatchState_Uncomplete {
+					continue
+				}
+
+				nodeID, err := parseNodeKey(key)
+				if err != nil {
+					log.Warn("fail to parse node from key", zap.String("key", key), zap.Error(err))
+					continue
+				}
+
+				ackEvent := parseAckEvent(nodeID, watchInfo)
+				c.processAck(ackEvent)
+			}
+		}
+	}
+}
+
+// Release writes ToRlease channel watch states for a channel
+func (c *ChannelManager) Release(nodeID UniqueID, channelName string, callback func(string) error) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.dropCallback == nil && callback != nil {
+		c.dropCallback = callback
```

Comment:
what if server crash?
How do we know if c.dropCallback exists or not?

Reply:
This callback is set at s.DropVChannel when calling `channelManager.Release(xxx, xxx, callback)`, and only when channel's collections is dropped, it'll need a callback later.